tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,705 @@
+
+        Ok(akes)
+    }
+
+    fn setup_tests() -> TestContext {
+        let ssh_dir_path: PathBuf = TempDir::new(SSH_DIR_PREFIX).unwrap().into_path();
+
+        for (name, text) in &*TEST_KEYS {
+            let pub_path = format!("{}/{}.pub", ssh_dir_path.to_str().unwrap(), name);
+
+            let mut pubfile = File::create(pub_path.clone()).expect(format!("unable to create a file {}", pub_path.as_str()).as_str());
+            let _ = pubfile.write_all(text.as_bytes());
+        }
+
+        TestContext {
+            ssh_dir: ssh_dir_path,
+        }
+    }
+
+    #[test]
+    fn test_no_keys() {
+        let ctx = setup_tests();
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
+
+        let aks = open_authorized_keys(ssh_dir);
+
+        aks.write().expect("failed to update authorized keys directory");
+        assert!(format!("{}", aks.sync().unwrap_err().kind()).contains("no keys found"));
+
+        let authkeysdir: PathBuf = PathBuf::from(format!("{}/authorized_keys.d", ssh_dir));
+        assert!(authkeysdir.is_dir());
+
+        let authkeys: PathBuf = PathBuf::from(format!("{}/authorized_keys", ssh_dir));
+        assert!(!authkeys.exists());
+    }
+
+    #[test]
+    fn test_first_run() {
+        let ctx = setup_tests();
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
+
+        let authkeys: PathBuf = PathBuf::from(format!("{}/authorized_keys", ssh_dir));
+        let mut authkeysfile =
+            File::options().create(true).append(true).write(true).open(authkeys.clone()).expect(format!("unable to create a file {}", authkeys.to_str().unwrap_or_default()).as_str());
+        for (_, text) in &*TEST_KEYS {
+            _ = authkeysfile.write_all(format!("{}\n", text).as_bytes());
+        }
+
+        // equivalent of running "update-ssh-keys" without args
+        let aks = open_authorized_keys(ssh_dir);
+
+        aks.write().expect("failed to update authorized keys directory");
+        aks.sync().expect("failed to update authorized keys");
+
+        let authkeys: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/old_authorized_keys", ssh_dir));
+        assert!(authkeys.exists());
+        assert_has_keys(ssh_dir, ["valid1", "valid2", "valid3"].to_vec(), true);
+    }
+
+    #[test]
+    fn test_add_one_file() {
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
+        let pubkeyvalid1 = "valid1";
+        let testnameone = "one";
+        let assert_keys = ["valid1"].to_vec();
+
+        // "update-ssh-keys --add one valid1.pub"
+        match add_one_ssh_key(
+            ssh_dir,      // ssh_dir
+            pubkeyvalid1, // pubkeyname
+            testnameone,  // testname
+            false,        // is_force
+            true,         // is_replace
+        ) {
+            Ok(keys) => add_key_check_results(ssh_dir, keys, pubkeyvalid1, testnameone, assert_keys),
+            Err(err) => panic!("update_ssh_keys --add failed {}", err),
+        }
+    }
```

Comment:
The test case misses the original the final assertion from Python:
```
 134   │         self.assertHasKeys('valid1')
```

I.e.: `assert_has_keys(ssh_dir, ["valid1"].to_vec(), true);`

Reply:
Ah, now I see `add_key_check_results` chains a call to `assert_has_keys`